src/main/java/top/yatt/dfargx/automata/DFA.java

```diff
@@ -40,6 +40,33 @@ public int getInitState() {
         return fs;
     }
 
+    @Override
+    public boolean equals(Object o) {
+        if (!(o instanceof DFA)) {
+            return false;
+        }
+        DFA other = (DFA) o;
+        if (other.is == this.is && other.rs == this.rs) {

```

Comment:
I guess this is based on the assumption that initial state and rejected state are always the same after compilation of "different but equal" regex patterns.

Is this assumption always working?

Reply:
Correct. If I understood correctly, a regular expression contains another if and only if their DFA's are _identical_, which means all their attributes are (including initial state and rejected state).

I'm sure your knowledge about DFA is far more deep than mine, so please correct me if I'm wrong!

Reply:
Yes, the equal regex pattern should have identical DFA, but the identical DFAs can have different state numbers.

e.g.
### DFA1

|            |a   |b|
|-----------|----|---|
|1(initial) | 2   |3|
|2(accepted) |3   |3|
|3(rejected) |3   |3|

### DFA2
|            |a   |b|
|-----------|----|---|
|1(rejected) | 1  |1|
|2(initial) |3   |1|
|3(accepted) |1  |1|

These two DFAs are exactly the same and minimal, but still have state numbers different.

Maybe we can make the "equals" method cover this case?

Reply:
Ok, I get what you mean, and I wonder why tests are just working fine...
I could try to write a better equals method, but unfortunately I don't have the required knowledge.

Do you have a good link to some documentation example on DFA equivalence?